File: akka-docs-new/src/test/scala/tutorial_1/ActorHierarchyExperiments.scala
package tutorial_1

import akka.actor.{ Actor, Props }
import akka.testkit.AkkaSpec

//#print-refs
class PrintMyActorRefActor extends Actor {
  override def receive: Receive = {
    case "printit" =>
      val secondRef = context.actorOf(Props.empty, "second-actor")
      println(secondRef)
  }
}
//#print-refs

//#start-stop
class StartStopActor1 extends Actor {
  override def receive: Receive = {
    case "stop" => context.stop(self)
  }

  override def preStart(): Unit = {
    println("first started")
    context.actorOf(Props[StartStopActor2], "second")
  }
  override def postStop(): Unit = println("first stopped")
}

class StartStopActor2 extends Actor {
  // Actor.emptyBehavior is a useful placeholder when we don't
  // want to handle any messages in the actor.
  override def receive: Receive = Actor.emptyBehavior

  override def preStart(): Unit = println("second started")
  override def postStop(): Unit = println("second stopped")
}
//#start-stop

//#supervise
class SupervisingActor extends Actor {
  val child = context.actorOf(Props[SupervisedActor], "supervised-actor")

  override def receive: Receive = {
    case "failChild" => child ! "fail"
  }
}

class SupervisedActor extends Actor {

  override def preStart(): Unit = println("supervised actor started")
  override def postStop(): Unit = println("supervised actor stopped")

  override def receive: Receive = {
    case "fail" =>
      println("supervised actor fails now")
      throw new Exception("I failed!")
  }
}
//#supervise

class ActorHierarchyExperiments extends AkkaSpec {

  "create top and child actor" in {

    //#print-refs
    val firstRef = system.actorOf(Props[PrintMyActorRefActor], "first-actor")
    println(firstRef)
    firstRef ! "printit"
    //#print-refs

  }

  "start and stop actors" in {
    //#start-stop
    val first = system.actorOf(Props[StartStopActor1], "first")
    first ! "stop"
    //#start-stop
  }

  "supervise actors" in {
    //#supervise
    val supervisingActor = system.actorOf(Props[SupervisingActor], "supervising-actor")
    supervisingActor ! "failChild"
    //#supervise
  }

}

File: akka-docs/src/test/java/jdocs/tutorial_1/ActorHierarchyExperiments.java
/*
 * Copyright (C) 2018-2019 Lightbend Inc. <https://www.lightbend.com>
 */

// #print-refs
package com.example;

// #print-refs

import akka.testkit.javadsl.TestKit;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;
import org.scalatestplus.junit.JUnitSuite;

// #print-refs
import akka.actor.AbstractActor;
import akka.actor.AbstractActor.Receive;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;

class PrintMyActorRefActor extends AbstractActor {
  static Props props() {
    return Props.create(PrintMyActorRefActor.class, PrintMyActorRefActor::new);
  }

  @Override
  public Receive createReceive() {
    return receiveBuilder()
        .matchEquals(
            "printit",
            p -> {
              ActorRef secondRef = getContext().actorOf(Props.empty(), "second-actor");
              System.out.println("Second: " + secondRef);
            })
        .build();
  }
}
// #print-refs

// #start-stop
class StartStopActor1 extends AbstractActor {
  static Props props() {
    return Props.create(StartStopActor1.class, StartStopActor1::new);
  }

  @Override
  public void preStart() {
    System.out.println("first started");
    getContext().actorOf(StartStopActor2.props(), "second");
  }

  @Override
  public void postStop() {
    System.out.println("first stopped");
  }

  @Override
  public Receive createReceive() {
    return receiveBuilder()
        .matchEquals(
            "stop",
            s -> {
              getContext().stop(getSelf());
            })
        .build();
  }
}

class StartStopActor2 extends AbstractActor {

  static Props props() {
    return Props.create(StartStopActor2.class, StartStopActor2::new);
  }

  @Override
  public void preStart() {
    System.out.println("second started");
  }

  @Override
  public void postStop() {
    System.out.println("second stopped");
  }

  // Actor.emptyBehavior is a useful placeholder when we don't
  // want to handle any messages in the actor.
  @Override
  public Receive createReceive() {
    return receiveBuilder().build();
  }
}
// #start-stop

// #supervise
class SupervisingActor extends AbstractActor {
  static Props props() {
    return Props.create(SupervisingActor.class, SupervisingActor::new);
  }

  ActorRef child = getContext().actorOf(SupervisedActor.props(), "supervised-actor");

  @Override
  public Receive createReceive() {
    return receiveBuilder()
        .matchEquals(
            "failChild",
            f -> {
              child.tell("fail", getSelf());
            })
        .build();
  }
}

class SupervisedActor extends AbstractActor {
  static Props props() {
    return Props.create(SupervisedActor.class, SupervisedActor::new);
  }

  @Override
  public void preStart() {
    System.out.println("supervised actor started");
  }

  @Override
  public void postStop() {
    System.out.println("supervised actor stopped");
  }

  @Override
  public Receive createReceive() {
    return receiveBuilder()
        .matchEquals(
            "fail",
            f -> {
              System.out.println("supervised actor fails now");
              throw new Exception("I failed!");
            })
        .build();
  }
}
// #supervise

// #print-refs
public class ActorHierarchyExperiments {
  public static void main(String[] args) throws java.io.IOException {
    ActorSystem system = ActorSystem.create("testSystem");

    ActorRef firstRef = system.actorOf(PrintMyActorRefActor.props(), "first-actor");
    System.out.println("First: " + firstRef);
    firstRef.tell("printit", ActorRef.noSender());

    System.out.println(">>> Press ENTER to exit <<<");
    try {
      System.in.read();
    } finally {
      system.terminate();
    }
  }
}
// #print-refs

class ActorHierarchyExperimentsTest extends JUnitSuite {
  static ActorSystem system;

  @BeforeClass
  public static void setup() {
    system = ActorSystem.create();
  }

  @AfterClass
  public static void teardown() {
    TestKit.shutdownActorSystem(system);
    system = null;
  }

  @Test
  public void testStartAndStopActors() {
    // #start-stop-main
    ActorRef first = system.actorOf(StartStopActor1.props(), "first");
    first.tell("stop", ActorRef.noSender());
    // #start-stop-main
  }

  @Test
  public void testSuperviseActors() {
    // #supervise-main
    ActorRef supervisingActor = system.actorOf(SupervisingActor.props(), "supervising-actor");
    supervisingActor.tell("failChild", ActorRef.noSender());
    // #supervise-main
  }
}
